### DISCUSSION TITLE
Return a connection object from a function?

### DISCUSSION THREAD

**Discussion by Ziris85** [diwic/dbus-rs#343] [Q&A] [answered]
Hello all,

I'm fairly new to rust and dbus (working through various tutorials and walkthroughs as I go along), and I've written up this bit of code that won't compile due to ownership/lifetime issues, and I was hoping someone here might be able to point me in the right direction on how to overcome:
```
enum DbusInterface {
    Timedate,
    Systemd,
}

impl DbusInterface {
    
    fn connect(&self) -> Result<Proxy<'_, &dbus::blocking::Connection>, dbus::Error> {
        let c = Connection::new_system()?;
       match &self {
            DbusInterface::Timedate => Ok{c.with_proxy("org.freedesktop.timedate1", "/org/freedesktop/timedate1", Duration::from_millis(5000))),
            DbusInterface::Systemd => Ok(c.with_proxy("org.freedesktop.systemd", "/org/freedesktop/systemd", Duration::from_millis(5000))),
        }
    }
//    fn get() {}
//    fn get_all() {}

}
```
The end goal is to be able to return the connection to be used in other functions to interact with dbus in various ways. However, Rust won't compile this with:
```
error[E0515]: cannot return value referencing local variable `c`
  --> src/main.rs:18:40
   |
18 | ...imedate => Ok(c.with_proxy("org.freedesktop.timedate1", "/org/freedesktop/timedate1", Duration::from_millis(5000))),
   |               ^^^-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |               |  |
   |               |  `c` is borrowed here
   |               returns a value referencing data owned by the current function
```
Attempting to dereference also just gives me an "lol no" error message. I'm sure I'm missing something obvious here, and perhaps I'm not asking in the right location (maybe stackoverflow?), but I thought I'd start here. I would be greatly appreciative of any guidance. Thank you!

**Comment by diwic** [comment, marked as answer]
I think there isn't one right answer to this question; there are several designs. Probably the easiest one is to create a struct like this:

    struct MyConnection {
        conn: Connection,
        intf: DBusInterface // I e, your existing enum
    }

And then in every method of `MyConnection`, construct the proxy and then call whatever methods you need on that.

The other option would be to create a struct that contains both a `Connection` and a `Proxy` and that struct then has to use a crate like `rental` because the `Proxy` references the `Connection`. I guess this is more complicated, but I don't know because I never used `rental` myself.

**Reply by Ziris85** [reply]
diwic,

Thanks for the reply! I'm not sure that accomplishes what I was hoping to do here, since my main goal was to reduce the amount of redundant code. For example, if I set the Timedate enum, then I'll always know what the name of my interface is, and I was looking to establish that proxy in one method and pass that to another method to do stuff with. However, as it seems that the proxy may not be able to be passed from one method to another, I'll have to build the connection individually in each method. So for example, instead of:
```
fn connect(whatever) -> whatever {
    // Do stuff to establish connection, such as my earlier example
}
fn get(whatever) -> whatever {
    let c = connect();
    // do stuff with connection
}
```
I'll have to just, in each method:
```
fn get(whatever) -> whatever {
    let c = Connection::new_system();
    let p = c.with_proxy(stuff);
    // do stuff with connection
}
```
Probably with some match arms interspersed in there as well. If that's the case, I'll just have to find a way to live with it. Thanks again for your time and help!

**Reply by diwic** [reply]
Actually I have a better suggestion, that I didn't think of when I wrote my earlier answer. Just construct the Proxy like this instead:

     Proxy::new("some.dest", "/some/path", Duration::from_millis(5000), Connection::new_system())
 

Now you can return it :-)

**Reply by Ziris85** [reply]
diwic,

Thanks again for the reply! I gave your new suggestion a try as-is and it complained about the Result value being present instead:
```
DBusInterface::Timedate => Proxy::new("org.freedesktop.timedate1", "/org/freedesktop/timedate1", Duration::from_millis(5000), Connection::new_system()),

error[E0308]: mismatched types
  --> src/main.rs:71:137
   |
71 | ...rom_millis(5000), Connection::new_system())),
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^ expected `&dbus::blocking::Connection`, found enum `std::result::Result`
   |
   = note: expected reference `&dbus::blocking::Connection`
                   found enum `std::result::Result<dbus::blocking::Connection, dbus::Error>`
```

Adding a simple unwrap to that gets to something else that was hinted at earlier - that it expects a borrowed value to be present there:
```
DBusInterface::Timedate => Proxy::new("org.freedesktop.timedate1", "/org/freedesktop/timedate1", Duration::from_millis(5000), Connection::new_system().unwrap()),

error[E0308]: mismatched types
  --> src/main.rs:70:139
   |
70 | ...s(5000), Connection::new_system().unwrap()),
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |             |
   |             expected `&dbus::blocking::Connection`, found struct `dbus::blocking::Connection`
   |             help: consider borrowing here: `&Connection::new_system().unwrap()`
```

And of course, making that a borrowed value takes us back to the original problem:
```
DBusInterface::Timedate => Proxy::new("org.freedesktop.timedate1", "/org/freedesktop/timedate1", Duration::from_millis(5000), &Connection::new_system().unwrap()),

error[E0515]: cannot return value referencing temporary value
  --> src/main.rs:70:40
   |
70 | ...=> Proxy::new("org.freedesktop.timedate1", "/org/freedesktop/timedate1", Duration::from_millis(5000), &Connection::new_system().unwrap(...
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^---------------------------------^
   |       |                                                                                                   |
   |       |                                                                                                   temporary value created here
   |       returns a value referencing data owned by the current function
```
I appreciate any other ideas or suggestions you might have here! I did manage to come up with a solution that involves a messy struct and such, but being able to do something like what you mentioned would be FAR cleaner and simpler code indeed :smile:

**Reply by diwic** [reply]
Make sure there is no `&` in the return value of the function, it should look like this:

    fn connect(&self) -> Result<Proxy<'static, dbus::blocking::Connection>, dbus::Error>

Now you can add a `?` after `new_system()` to have it handle errors more gracefully.

**Reply by Ziris85** [reply]
Ah yes, thank you! That gets me a bit closer to making it work. It does successfully allow me to return that value now, but I'm having a hard time actually doing anything with it due to trait bounds. Here's what I have at this point (forgive the poor quality of the code):
```
pub enum DBusInterface {
    Timedate,
}

struct DBusRequest<'a> {
    conn: Proxy<'a, dbus::blocking::Connection>,
    url: String
}

impl DBusInterface {
    fn connect(&self) -> Result<DBusRequest, dbus::Error> {
        match &self {
            DBusInterface::Timedate => Ok(
                DBusRequest {
                    conn: Proxy::new("org.freedesktop.timedate1", "/org/freedesktop/timedate1", Duration::from_millis(5000), Connection::new_system()?),
                    url: String::from("org.freedesktop.timedate1")
                } 
            )
        }
    }
    pub fn get(&self, req: String) -> Box<dyn arg::RefArg> {
        let p = self.connect().unwrap();
        p.conn.get(p.url, &req).unwrap()
    }
}
```
Compiling returns:
```
error[E0599]: no method named `get` found for struct `Proxy<'_, dbus::blocking::Connection>` in the current scope
   --> src/lib.rs:94:16
    |
94  |         p.conn.get(p.url, &req).unwrap()
    |                ^^^ method not found in `Proxy<'_, dbus::blocking::Connection>`
    | 
   ::: /root/.cargo/registry/src/github.com-1ecc6299db9ec823/dbus-0.9.3/src/blocking.rs:117:1
    |
117 | pub struct Connection {
    | ---------------------
    | |
    | doesn't satisfy `<dbus::blocking::Connection as Deref>::Target = _`
    | doesn't satisfy `dbus::blocking::Connection: Deref`
...
315 | pub struct Proxy<'a, C> {
    | ----------------------- doesn't satisfy `_: dbus::blocking::stdintf::org_freedesktop_dbus::Properties`
    |
    = note: the method `get` exists but the following trait bounds were not satisfied:
            `<dbus::blocking::Connection as Deref>::Target = _`
            which is required by `Proxy<'_, dbus::blocking::Connection>: dbus::blocking::stdintf::org_freedesktop_dbus::Properties`
            `dbus::blocking::Connection: Deref`
            which is required by `Proxy<'_, dbus::blocking::Connection>: dbus::blocking::stdintf::org_freedesktop_dbus::Properties`
```
I'm a little confused as to why this is happening, since as I understand it, the "with_proxy" method is just a convenience wrapper that returns a Proxy struct that I'm just constructing by hand, so I'm not quite sure how/why this is different. Traits are still something of a confusing topic for me, so that's probably why I'm struggling here. I've read through the Rust docs on them and they seem really neat/useful! Just not quite sure how to work with them yet heh.

**Reply by diwic** [reply]
Hmm, so Deref is not implemented for itself, sorry. Easiest (but slightly annoying) is to just wrap it in a box:

    struct DBusRequest {
        conn: Proxy<'static, Box<dbus::blocking::Connection>>,
        url: String
    }

And when creating the proxy:

    Proxy::new("org.freedesktop.timedate1", "/org/freedesktop/timedate1", Duration::from_millis(5000), Box::new(Connection::new_system()?))

**Reply by Ziris85** [reply]
I forgot to follow up on this! This worked beautifully, thank you for your help!
